File: examples/00-quickstart/python/quickstart.py
"""
Jigyasa Quickstart — 5 minutes from zero to search.

Prerequisites:
    pip install grpcio grpcio-tools grpcio-reflection googleapis-common-protos

Usage:
    python quickstart.py
"""
import grpc
import json
import subprocess
import sys
import os
import time

ADDR = "localhost:50051"
PROTO_DIR = os.path.join(os.path.dirname(__file__), "..", "..", "..", "src", "main", "proto")
GEN_DIR = os.path.join(os.path.dirname(__file__), "_gen")


def generate_stubs():
    """Generate Python gRPC stubs from proto."""
    os.makedirs(GEN_DIR, exist_ok=True)
    import grpc_tools
    grpc_include = os.path.join(os.path.dirname(grpc_tools.__file__), "_proto")
    import site
    site_pkgs = [p for p in site.getsitepackages() if "site-packages" in p][0]

    result = subprocess.run([
        sys.executable, "-m", "grpc_tools.protoc",
        f"--proto_path={PROTO_DIR}",
        f"--proto_path={grpc_include}",
        f"--proto_path={site_pkgs}",
        f"--python_out={GEN_DIR}",
        f"--grpc_python_out={GEN_DIR}",
        "dpSearch.proto"
    ], capture_output=True, text=True)
    if result.returncode != 0:
        print(f"Proto compilation failed: {result.stderr}")
        sys.exit(1)
    sys.path.insert(0, GEN_DIR)


def main():
    generate_stubs()
    import dpSearch_pb2 as pb
    import dpSearch_pb2_grpc as pb_grpc

    channel = grpc.insecure_channel(ADDR)
    stub = pb_grpc.JigyasaDataPlaneServiceStub(channel)

    # 1. Health check
    print("1. Health check")
    health = stub.Health(pb.HealthRequest())
    print(f"   Status: {pb.HealthResponse.Status.Name(health.status)}")

    # 2. Create collection
    print("\n2. Create collection 'quickstart'")
    schema = json.dumps({"fields": [
        {"name": "id", "type": "STRING", "key": True, "filterable": True},
        {"name": "title", "type": "STRING", "searchable": True, "sortable": True},
        {"name": "body", "type": "STRING", "searchable": True},
        {"name": "category", "type": "STRING", "filterable": True},
    ]})
    try:
        stub.CreateCollection(pb.CreateCollectionRequest(
            collection="quickstart", indexSchema=schema))
        print("   ✓ Created")
    except grpc.RpcError as e:
        if "already exists" in str(e.details()):
            print("   ✓ Already exists")
        else:
            raise

    # 3. Index documents
    print("\n3. Index 5 documents")
    docs = [
        {"id": "1", "title": "Getting Started with Lucene", "body": "Apache Lucene is a high-performance search engine library written in Java", "category": "search"},
        {"id": "2", "title": "Vector Search Fundamentals", "body": "HNSW algorithm enables fast approximate nearest neighbor search", "category": "search"},
        {"id": "3", "title": "Building AI Agents", "body": "Autonomous agents use LLMs to reason, plan, and take actions", "category": "ai"},
        {"id": "4", "title": "gRPC Protocol Guide", "body": "gRPC uses HTTP/2 and Protocol Buffers for efficient communication", "category": "networking"},
        {"id": "5", "title": "Memory for LLM Agents", "body": "Agents need persistent memory to recall past conversations and facts", "category": "ai"},
    ]
    items = [pb.IndexItem(document=json.dumps(d)) for d in docs]
    resp = stub.Index(pb.IndexRequest(item=items, collection="quickstart"))
    print(f"   ✓ Indexed {len(resp.itemResponse)} documents")

    time.sleep(1)  # Wait for NRT refresh

    # 4. Search
    print("\n4. Search for 'search engine library'")
    resp = stub.Query(pb.QueryRequest(
        collection="quickstart",
        text_query="search engine library",
        include_source=True,
        top_k=5
    ))
    print(f"   Total hits: {resp.total_hits}")
    for hit in resp.hits:
        doc = json.loads(hit.source)
        print(f"   [{hit.score:.3f}] {doc['title']}")

    # 5. Count
    print(f"\n5. Total documents in collection")
    count = stub.Count(pb.CountRequest(collection="quickstart"))
    print(f"   Count: {count.count}")

    print("\n✅ Quickstart complete!")
    channel.close()


if __name__ == "__main__":
    main()

File: examples/00-quickstart/java/src/main/java/quickstart/Quickstart.java
package quickstart;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import com.jigyasa.dp.search.protocol.JigyasaDataPlaneServiceGrpc;
import com.jigyasa.dp.search.protocol.JigyasaDataPlaneServiceGrpc.JigyasaDataPlaneServiceBlockingStub;
import com.jigyasa.dp.search.protocol.*;

import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * Jigyasa Quickstart — 5 minutes from zero to search.
 *
 * Usage: gradle run
 */
public class Quickstart {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    public static void main(String[] args) throws Exception {
        ManagedChannel channel = ManagedChannelBuilder
                .forAddress("localhost", 50051)
                .usePlaintext()
                .build();

        JigyasaDataPlaneServiceBlockingStub stub =
                JigyasaDataPlaneServiceGrpc.newBlockingStub(channel);

        // 1. Health check
        System.out.println("1. Health check");
        HealthResponse health = stub.health(HealthRequest.getDefaultInstance());
        System.out.println("   Status: " + health.getStatus());

        // 2. Create collection
        System.out.println("\n2. Create collection 'quickstart-java'");
        String schema = MAPPER.writeValueAsString(MAPPER.createObjectNode().set("fields",
                MAPPER.createArrayNode()
                        .add(field("id", "STRING", true, false, true, false))
                        .add(field("title", "STRING", false, true, false, true))
                        .add(field("body", "STRING", false, true, false, false))
                        .add(field("category", "STRING", false, false, true, false))
        ));
        try {
            stub.createCollection(CreateCollectionRequest.newBuilder()
                    .setCollection("quickstart-java")
                    .setIndexSchema(schema)
                    .build());
            System.out.println("   ✓ Created");
        } catch (io.grpc.StatusRuntimeException e) {
            if (e.getMessage().contains("already exists")) {
                System.out.println("   ✓ Already exists");
            } else {
                throw e;
            }
        }

        // 3. Index documents
        System.out.println("\n3. Index 5 documents");
        List<IndexItem> items = List.of(
                doc("1", "Getting Started with Lucene", "Apache Lucene is a high-performance search engine library written in Java", "search"),
                doc("2", "Vector Search Fundamentals", "HNSW algorithm enables fast approximate nearest neighbor search", "search"),
                doc("3", "Building AI Agents", "Autonomous agents use LLMs to reason, plan, and take actions", "ai"),
                doc("4", "gRPC Protocol Guide", "gRPC uses HTTP/2 and Protocol Buffers for efficient communication", "networking"),
                doc("5", "Memory for LLM Agents", "Agents need persistent memory to recall past conversations and facts", "ai")
        );
        IndexResponse indexResp = stub.index(IndexRequest.newBuilder()
                .setCollection("quickstart-java")
                .addAllItem(items)
                .build());
        System.out.println("   ✓ Indexed " + indexResp.getItemResponseCount() + " documents");

        Thread.sleep(1000); // Wait for NRT refresh

        // 4. Search
        System.out.println("\n4. Search for 'search engine library'");
        QueryResponse queryResp = stub.query(QueryRequest.newBuilder()
                .setCollection("quickstart-java")
                .setTextQuery("search engine library")
                .setIncludeSource(true)
                .setTopK(5)
                .build());
        System.out.println("   Total hits: " + queryResp.getTotalHits());
        for (QueryHit hit : queryResp.getHitsList()) {
            var doc = MAPPER.readTree(hit.getSource());
            System.out.printf("   [%.3f] %s%n", hit.getScore(), doc.get("title").asText());
        }

        // 5. Count
        System.out.println("\n5. Total documents in collection");
        CountResponse count = stub.count(CountRequest.newBuilder()
                .setCollection("quickstart-java").build());
        System.out.println("   Count: " + count.getCount());

        System.out.println("\n✅ Quickstart complete!");
        channel.shutdown().awaitTermination(5, TimeUnit.SECONDS);
    }

    private static ObjectNode field(String name, String type, boolean key,
                                     boolean searchable, boolean filterable, boolean sortable) {
        ObjectNode f = MAPPER.createObjectNode();
        f.put("name", name);
        f.put("type", type);
        if (key) f.put("key", true);
        if (searchable) f.put("searchable", true);
        if (filterable) f.put("filterable", true);
        if (sortable) f.put("sortable", true);
        return f;
    }

    private static IndexItem doc(String id, String title, String body, String category) {
        try {
            ObjectNode node = MAPPER.createObjectNode();
            node.put("id", id);
            node.put("title", title);
            node.put("body", body);
            node.put("category", category);
            return IndexItem.newBuilder()
                    .setDocument(MAPPER.writeValueAsString(node))
                    .build();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}
